lib/core/src/math/utils.rs

```diff
@@ -56,3 +71,8 @@ pub fn invert_y(mut m: Matrix4<f32>) -> Matrix4<f32> {
     m.row_part_mut(1, 4).iter_mut().for_each(|a| *a = -*a);
     m
 }
+
+// Note [2D array to 1D array]
+// ===========================
+// We need to interpret [[f32; 4]; 4] as [f32; 16] to view as Float32Array.
+// The memory layout is the same, so this operation is safe.
```

Comment:
Not quite. The `Float32Array` from `view` may be invalidated if there is _any_ allocation done after its creation. So if someone not suspecting anything (because function is safe, right?) creates Float32Array from matrix and makes a simple Box before reading array, he may be surprised

Reply:
And you assumes that Matrix4 from nalgebra keeps data in `[[f32; 4]; 4]` what might be not the case, in fact i don't see in nalgebra docs any guarantees about this. So the code is very unstable, a upgrade of nalgebra may cause unexpected behaviours (or new optimization in rustc)

Reply:
@farmaazon I am not sure I understood the scenario. What kind of allocation can break it? Can you provide more detail, maybe with an example?

About the memory layout, here is the documentation for AsRef:

https://docs.rs/nalgebra/0.19.0/nalgebra/base/struct.Matrix.html#impl-AsRef%3C%5B%5BN%3B%204%5D%3B%204%5D%3E

This kind of memory layout for matrices is common in math libraries. If the Matrix memory layout (including `majorness`) breaks in future releases, we can catch it with our units tests.

I have added this test to make sure we have the expected memory layout:

```rust
    #[test]
    fn matrix4_memory_layout() {
        use super::Matrix4;
        let matrix = Matrix4::<f32>::new
            ( 1.0, 5.0,  9.0, 13.0
            , 2.0, 6.0, 10.0, 14.0
            , 3.0, 7.0, 11.0, 15.0
            , 4.0, 8.0, 12.0, 16.0 );
        let matrix = matrix.as_ref();
        unsafe {
            let layout = &*(matrix as *const [[f32; 4]; 4] as *const [f32; 16]);
            for (i, n) in layout.iter().enumerate() {
                assert_eq!((i + 1) as f32, *n);
            }
        }
    }
```

Reply:
Also, this can make this function safe:
```rust
#[derive(Shrinkwrap)]
pub struct Float32ArrayView<'a> {
    #[shrinkwrap(main_field)]
    array   : Float32Array,
    phantom : PhantomData<&'a Float32Array>
}

pub trait IntoFloat32Array {
    fn into_float32array<'a>(&'a self) -> Float32ArrayView<'a>;
}

impl IntoFloat32Array for Matrix4<f32> {
    fn into_float32array<'a>(&'a self) -> Float32ArrayView<'a> {
        // Note [2D array to 1D array]
        unsafe {
            let matrix = self.as_ref();
            let matrix = matrix as *const [[f32; 4]; 4];
            let array = Float32Array::view(&*(matrix as *const [f32; 16]));
            let phantom = PhantomData;
            Float32ArrayView { array, phantom }
        }
    }
}
```

Reply:
@farmaazon is right. This is **VERY** unsafe and we should not use it. See the docs of `js_sys::Float32Array::view`: 
```
Views into WebAssembly memory are only valid so long as the backing buffer isn't resized in JS. Once this function is called any future calls to Box::new (or malloc of any form) may cause the returned value here to be invalidated. Use with caution!

Additionally the returned object can be safely mutated but the input slice isn't guaranteed to be mutable.

Finally, the returned object is disconnected from the input slice's lifetime, so there's no guarantee that the data is read at the right time.
```

Keeping this `View` in a structure is a very, very, very bad idea. Danilo, I've got a request - when writing docs about safety, please address **every** safety note from **every** unsafe function you use there, including for example the `view` function. 

Btw, I think we should introduce the same thing the Firefox team did:

Reply:
![tall_unsafe_1](https://user-images.githubusercontent.com/1623053/69853526-71f52e80-1287-11ea-820d-f8a07ae3b920.jpg)

Reply:
![tall_unsafe_2](https://user-images.githubusercontent.com/1623053/69853541-7c172d00-1287-11ea-83e1-21334a408c5d.jpg)

Reply:
With AsRef you're right, that's legitimate use of casting.
As for `Flotat32Array::view` the example is:
```rust
let arr = matrix.into_float32array();
let some_data : Box<Whatever> = Box::new(); // here is new allocation, 
// so created view is invalidated (see a docs fragment mentioned by @wdanilo )
let_pass_this_array_to_js(arr); // however an user does not expect it 
// and pass invalid view somewhere else

Reply:
Thanks for bringing that to my attention. I missed the Unsafety note of `Float32Array::view`.

I qualified the function as unsafe and I also provided documentation pointing out why it's unsafe.

Reply:
I don't see it. The code is still:

```rust
impl IntoFloat32Array for Matrix4<f32> {
    fn into_float32array(&self) -> Float32Array {
        // Note [2D array to 1D array]
        unsafe {
            let matrix = self.as_ref();
            Float32Array::view(&*(matrix as *const [[f32; 4]; 4] as *const [f32; 16]))
        }
    }
}
```

Reply:
Btw, let's move the note to the doc comment of the whole function

Reply:
@wdanilo I pushed it few minutes ago. Please check it now.